Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp

```diff
@@ -0,0 +1,505 @@
+    });
+    return true;
+}
+
+// https://tc39.es/proposal-temporal/#prod-Hour
+bool ISO8601Parser::parse_hour()
+{
+    // Hour :
+    //     0 DecimalDigit
+    //     1 DecimalDigit
+    //     20
+    //     21
+    //     22
+    //     23
+    TRY_PARSE({
+        if (m_state.lexer.consume_specific('0') || m_state.lexer.consume_specific('1')) {
+            PARSE_REQUIRED(parse_decimal_digit());
+        } else {
+            PARSE_REQUIRED(
+                m_state.lexer.consume_specific("20"sv)
+                || m_state.lexer.consume_specific("21"sv)
+                || m_state.lexer.consume_specific("22"sv)
+                || m_state.lexer.consume_specific("23"sv));
+        }
+    });
+    return true;
+}
+
+// https://tc39.es/proposal-temporal/#prod-MinuteSecond
+bool ISO8601Parser::parse_minute_second()
+{
+    // MinuteSecond :
+    //     0 DecimalDigit
+    //     1 DecimalDigit
+    //     2 DecimalDigit
+    //     3 DecimalDigit
+    //     4 DecimalDigit
+    //     5 DecimalDigit
+    TRY_PARSE({
+        PARSE_REQUIRED(
+            m_state.lexer.consume_specific('0')
+            || m_state.lexer.consume_specific('1')
+            || m_state.lexer.consume_specific('2')
+            || m_state.lexer.consume_specific('3')
+            || m_state.lexer.consume_specific('4')
+            || m_state.lexer.consume_specific('5'));
+        PARSE_REQUIRED(parse_decimal_digit());
+    });
+    return true;
+}
+
+// https://tc39.es/proposal-temporal/#prod-DecimalSeparator
+bool ISO8601Parser::parse_decimal_separator()
+{
+    // DecimalSeparator : one of
+    //     . ,
+    TRY_PARSE({
+        PARSE_REQUIRED(
+            m_state.lexer.consume_specific('.')
+            || m_state.lexer.consume_specific(','));
+    });
+    return true;
+}
+
+// https://tc39.es/proposal-temporal/#prod-DateTimeSeparator
+bool ISO8601Parser::parse_date_time_separator()
+{
+    // DateTimeSeparator :
+    //     <SP>
+    //     T
+    //     t
+    TRY_PARSE({
+        PARSE_REQUIRED(
+            m_state.lexer.consume_specific(' ')
+            || m_state.lexer.consume_specific('T')
+            || m_state.lexer.consume_specific('t'));
+    });
+    return true;
+}
+
+// https://tc39.es/proposal-temporal/#prod-DateYear
+bool ISO8601Parser::parse_date_year()
+{
+    // DateFourDigitYear :
+    //     DecimalDigit DecimalDigit DecimalDigit DecimalDigit
+    // DateExtendedYear :
+    //     Sign DecimalDigit DecimalDigit DecimalDigit DecimalDigit DecimalDigit DecimalDigit
+    // DateYear :
+    //     DateFourDigitYear
+    //     DateExtendedYear
+    m_state.parse_result.date_year = TRY_PARSE({
+        if (m_state.lexer.next_is(is_ascii_digit)) {
+            for (size_t i = 0; i < 4; ++i)
+                PARSE_REQUIRED(parse_decimal_digit());
+        } else {
+            PARSE_REQUIRED(parse_sign());
+            for (size_t i = 0; i < 6; ++i)
+                PARSE_REQUIRED(parse_decimal_digit());
+        }
+    });
+    return true;
```

Comment:
One more, using patterns that exist elsewhere in the codebase
```c++
    auto rule_state = current_state(); // RAII helper with a saved state, see below for a possible implementation

    if (parse_sign()) {
        for (size_t i = 0; i < 6; ++i) {
            if (!parse_decimal_digit()) return false;
        }
    } else {
        for (size_t i = 0; i < 4; ++i) {
            if (!parse_decimal_digit()) return false;
        }
    }

    m_state.parse_result.date_year =  m_input.substring_view(rule_state.start(), m_state.lexer.tell() - rule_state.start());
    rule_state.accept();
    return true;
```

Possible `current_state()` object
```c++
struct RuleState {
    RuleState(ISO8601Parser&, State) ...;

    ~RuleState()
    {
        if (!should_accept)
            m_parser.state = move(m_saved_state);
    }
    void accept() { m_should_accept = true; }

private:
    ISO8601Parser& m_parser;
    State m_saved_state;
    bool m_should_accept { false };
};
```

Reply:
Adopted this in a similar form, with credit :)